### PR TITLE
ADR 0009: Add network traffic diagrams

### PR DESCRIPTION
**Pull request by jameshochadel**
## Changes proposed in this pull request:
- Add diagram depicting before & after network traffic flow, to be used in SCR

## security considerations
None; documentation only

Note to self: Squash merge.

### CODE REVIEW

File: adr/ADR-0009-shield-advanced.md [hunk 1]
```diff
@@ -32,25 +32,59 @@ We already use two Managed Rulesets: Amazon IP Reputation List and Known Bad Inp
...
+        end
+        subgraph govcloud["AWS GovCloud"]
+            subgraph platform["cloud.gov Platform"]
+                appCF -->|App Request| appALB
+                mgmtCF -->|Mgmt Request| mgmtALB
+                mgmtCF -->|Auth Request| authALB
+
+                appALB[App ALB] -->|App Request| router[CF Router]
+                mgmtALB[Management ALB] -->|Mgmt Request| router
+                mgmtALB[Management ALB] -->|Auth Request| router
```
Comment (markdboyd) [inline adr/ADR-0009-shield-advanced.md:80]:
```suggestion
                authALB[Auth ALB] -->|Auth Request| router
```

I think the diagram is trying to show the separation of traffic for authentication to a separate LB, in which case i think the traffic should be shown going from the Auth ALB to the router.

Also, if we did create a new LB to handle traffic for UAA, I'm not sure if traffic would go through the CF gorouter?

Lastly, I assume it's intentional that we're showing the management and auth ALBs as protected by the same Cloudfront distribution. Would it be better to use separate distributions for the practical realities of clearing cache separately, etc?

**Comment by pburkholder** [comment]
Thanks for all the work on this, @jameshochadel. The diagrams in particular are really helpful.  

I'd like to check some of our assumptions, and make sure we document some of the options we're _not_ pursuing so we understand our reasoning later. 

One thing is that we seem to be attributing poor routing performance due to an increase in random malicious scans, but it also could be that work we've done in the last year, particularly around TLS-everywhere, could have impacted the efficient operation of our routers. I've seen other CF operators claim 2500 requests/second per router on a 2-core 4GB systems. We have past reports on our own system of 2500 req/s (probably across our entire system, not per router) back when we had fewer routers. 

I think we should look at 
- how we're using TLS - e.g. are we reusing TLS effectively between ALB and Routers, and Routers and Cells, or are we wasting a lot of resources in re-establishing those connections? Or are we burning resources using TLS from secureproxy to gorouter within localhost, instead using 127.0.0.1:8080?
- how did our load-testing with KR in January measure up to what we see today? Can we learn anything from that?
- do we have evidence that random-scanning has in fact increased by a significant factor? If not, then why are we doing this?

Another issue whether we're prepared for the complexity that may come with add'l CF distributions. With the "App Cloudfront distro" (I presumed for *.app.cloud.gov) will customers already using CloudFront  have requests going through two CloudFronts? Or will we need a separate origin for traffic already coming from CloudFront? Will we need to add AdvShield all existing CF distros? Will we still have useful visibility?

Third, are we convinced that Adv Shield is going to do what it need it to? Do we have the documentation and 3rd-party evaluation to validate that, or someway to validate it ourselves?  Do we have the ability to test this or do we just trust AWS on this?

Fourth, do we have any indication when Adv. Shield is coming to GovCloud? Some of their documentation is conflicting, like AWS Services-in-Scope page. Anyhow, it would seem a lot simpler to add Adv Shield our ALBs and not bother with addl CF distros, if we have the choice. Would we tear those down once Adv Shield is available in GovCloud?

Can we dig into some of the aformentioned before we give this the greene-light?

**Comment by jameshochadel** [comment]
@pburkholder

1. Addressing a few points:
  1. "One thing is that we seem to be attributing poor routing performance due to an increase in random malicious scans." I wouldn't say that. I am attributing system downtime to random malicious scans that exacerbate poor routing performance. Poor routing performance has some other underlying cause — perhaps one of the ones you state below.
  2. TLS: I am open to looking at this. That change was before my time, so I don't know much about it and can't speak to historical performance. Thank you for raising it!
  3. "Do we have evidence that random-scanning has in fact increased by a significant factor?" I have no evidence of an increase or decrease, but the rate of random scans isn't the motivation for the change; the impact when a scan occurs is.
2. Regarding "double CloudFront": I address this in writing the ADR, but I should add a diagram. No requests will go through CloudFront twice. The ALBs will be the origin for the new distributions and continue to be the origin for the existing distributions. This initial work would only protect customers who do not currently broker a CDN for themselves. Follow-on work would improve the External Domain Broker to protect brokered distributions as well.
3. IIRC we have positive feedback from the login team about Shield Advanced. I don't know of a good way to validate it.
4. Let's chat about this offline.

**Review by markdboyd — APPROVED** [review]
Approving given the ADR discussion yesterday. Great work! Especially love the diagrams, they're very helpful!

**Comment by jameshochadel** [comment]
Going to make a follow-up PR to integrate feedback from @pburkholder and clarify that we're only going to cover "platform" LBs, not the customer ones.

**Pull request merged**